### PR TITLE
Refactored fit on generator in Sequential model

### PR DESCRIPTION
**Pull request by jeanpijon**
Continuation of #1305 

Refactored fit on generator in Sequential model
Parallelization factored out from model to keras.utils.io_utils
Weakened asserts on data in model.Sequential.fit_on_generator - sometimes len(X) doesn't have to be same as len(y)

This is just the start of work, there would be modifications also in Graph model, but first I want to cleanup the Sequential model

### CODE REVIEW

File: keras/utils/io_utils.py [hunk 1]
```diff
@@ -69,3 +75,134 @@ def load_array(name):
     a[:] = array[:]
     f.close()
     return a
+
+def generate_infinit_iterators(iterator_constructor, *args, **kwargs):
```
Comment (olegsinyavskiy) [inline keras/utils/io_utils.py]: Hi!
Keras always uses 4 spaces as indentation

Comment (jeanpijon) [inline keras/utils/io_utils.py]: Good point, fixed

**Comment by olegsinyavskiy** [comment]
the reason you have so many pep8 errors is that you're not update with master of keras. there is a pytest.ini ignores that will make most of the errors disappear.

File: keras/models.py [hunk 2]
```diff
@@ -979,41 +908,72 @@ def generator_task():
...
-                            val_outs = self.evaluate(X, y,
-                                                     show_accuracy=show_accuracy,
-                                                     sample_weight=sample_weight,
-                                                     verbose=0)
-                        if type(val_outs) != list:
-                            val_outs = [val_outs]
-                        # same labels assumed
-                        for l, o in zip(out_labels, val_outs):
-                            epoch_logs['val_' + l] = o
+                if bool(samples_per_epoch) and samples_seen >= samples_per_epoch:  # epoch finished
```
Comment (gw0) [inline keras/models.py]: `if bool(samples_per_epoch)` is the same as `if samples_per_epoch`.

File: keras/models.py [hunk 3]
```diff
@@ -806,49 +777,59 @@ def load_weights(self, filepath):
...
+                which would create for each epoch new instance
+                X, y and sample_weight should be numpy array. len(y) and len(sample_weight) has to be the same
                 If the output tuple has two elements, they are assumed to be
                 (input_data, target_data).
                 If it has three elements, they are assumed to be
                 (input_data, target_data, sample_weight).
-                All arrays should contain the same number of samples.
+                If there are more than 3 items in tuple (e.g. metadata for scoring, names, etc.),they are ignored for training.
+
             samples_per_epoch: integer, number of samples to process before
```
Comment (gw0) [inline keras/models.py]: It should be noted that this is an optional epoch break point if your `data/epoch_generator` does not stop.

**Comment by gw0** [comment]
Interesting refactoring, but using generators of generators will likely confuse newbies and make the final code less readable. Or am I wrong?

What about having just one generator that outputs `None` or an `EpochEnd` object when the epoch ends?

**Comment by jeanpijon** [comment]
Well, I am also not really comfortable with this design, but to be honest, we were not able (so far) to come with any other reasonable solution. What do you mean by one generator design? Can you elaborate more on it?

**Comment by gw0** [comment]
With one-generator design I mean something like:

```
def data_gen(samples):
    while True:
        for sample in random.shuffle(samples):
            yield sample  # next sample
        yield None  # indicates epoch end
```

**Comment by jeanpijon** [comment]
Well, this is basically the same approach as I am using, the idea of generator over generators is this:
you have outer generator over epochs, and inner is over epoch data. (In a special case, you can have splitted iterator over infinit data, but all generalizes to the general form of genA(genB) )

**Comment by jeanpijon** [comment]
So I see, this was closed without comment and mroeover lots of work is going on in models.py.  Am I the only one, who would like to see more separation of paralelization code from the functional code? I am affraid, we are overcomplicating simple things, and moreover it seems to me, that incorporating parallelization is going straight against one of the ideal behind this:

> Minimalism. Each module should be kept short and simple. Every piece of code should be transparent upon first reading. No black magic: it hurts iteration speed and ability to innovate.

Do I miss something here?

**Pull request closed without merging**